# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Результаты исследования будут учтены при построении модели кредитного скоринга.

***Цель исследования*** — ответить на вопросы:

1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

***Ход исследования***

Входные данные от банка — статистика о платёжеспособности клиентов — получены из файла data.csv. О качестве данных ничего не известно. Поэтому перед тем, как отвечать на вопрсоы исследования и делать вывод, понадобится обзор данных.

Проверим данные на ошибки и оценим их влияние на исследование. Затем, на этапе предобработки, поищем возможность исправить самые критичные ошибки данных.

После ответа на поставленные в исследовании вопросы, сделаем общией вывод.

Таким образом, исследование пройдёт в четыре этапа:

1. Обзор данных
2. Предобработка данных
3. Ответы на вопросы исследования
4. Общий вывод

# Обзор данных

In [1]:
# импортируем библ и читаем файл
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [2]:
# Выведите первые 20 строчек датафрейма data
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [3]:
# Выведите основную информацию о датафрейме
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


# Предобработка данных

### Удаление пропусков

In [4]:
# Выведем количество пропущенных значений для каждого столбца
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Столбец с пропущенными значениями — total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type.

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

Обработаем пропуски в столбце days_employed. Отрицательное количество дней трудового стажа - это аномалия. Заменим все отрицательные значения положительными.

In [6]:
# возьмем модуль значений с помощью abs()
data['days_employed'] = data['days_employed'].abs()

In [7]:
# выведем медианное значение трудового стажа в днях
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится для исследования.

***Выведем перечень уникальных значений столбца children***

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.

In [9]:
# удаляем
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [10]:
# проверяем
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

***Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type.***

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [12]:
# убедимся, что все пропуски заполнены
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [13]:
# меняем вещественные значения на целочисленные
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [14]:
# приводим к нижнему регистру столбец education
data['education'] = data['education'].str.lower()

In [15]:
# вывод кол-ва строк дубликатов
data.duplicated().sum()

71

In [16]:
# удалим дубликаты
data = data.drop_duplicates()

### Категоризация данных

Создадим столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

***Выведем перечень уникальных целей взятия кредита***

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Сформирует новый столбец purpose_category, в который войдут следующие категории:

- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследование данных и ответы на вопросы

#### Для наглядности работы со столбцами отобразим их название

-	***children*** — количество детей в семье
-	***days_employed*** — общий трудовой стаж в днях
-	***dob_years*** — возраст клиента в годах
-	***education*** — уровень образования клиента
-	***education_id*** — идентификатор уровня образования
-	***family_status*** — семейное положение
-	***family_status_id*** — идентификатор семейного положения
-	***gender*** — пол клиента
-	***income_type*** — тип занятости
-	***debt*** — имел ли задолженность по возврату кредитов
-	***total_income*** — ежемесячный доход
-	***purpose*** — цель получения кредита

***Чтобы ответить на поставленные вопросы сделаем некоторые преобразования***

In [22]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

***Вывод: В столбце days_employed (общий трудовой стаж в днях) - вещественные значения. Должны быть целочисленные. Проверим и исправим этот недочет.***

In [23]:
data.info() # методом info проверим тип переменных в колонках

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21331 non-null  int64  
 1   days_employed          21331 non-null  float64
 2   dob_years              21331 non-null  int64  
 3   education              21331 non-null  object 
 4   education_id           21331 non-null  int64  
 5   family_status          21331 non-null  object 
 6   family_status_id       21331 non-null  int64  
 7   gender                 21331 non-null  object 
 8   income_type            21331 non-null  object 
 9   debt                   21331 non-null  int64  
 10  total_income           21331 non-null  int32  
 11  purpose                21331 non-null  object 
 12  total_income_category  21331 non-null  object 
 13  purpose_category       21331 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(7)
memory us

In [24]:
data['days_employed'] = data['days_employed'].astype(int) # с помощью метода astype() преобразуем данные к типу int
data.head() # проверка

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

***Вывод: Можно заметить, что в столбце days_employed присутствуют аномально большие значения. Посмотрим на эти значения.***

***Также посмотрим как обстоят дела с аномалиями в других столбцах***

In [25]:
# вывод показателей
data.describe()

children  days_employed     dob_years  education_id  \
count  21331.000000   21331.000000  21331.000000  21331.000000   
mean       0.474005   67148.470864     43.277952      0.816980   
std        0.752378  139281.136753     12.575300      0.549282   
min        0.000000      24.000000      0.000000      0.000000   
25%        0.000000    1024.000000     33.000000      1.000000   
50%        0.000000    1996.000000     42.000000      1.000000   
75%        1.000000    5337.000000     53.000000      1.000000   
max        5.000000  401755.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21331.000000  21331.000000  2.133100e+04  
mean           0.974497      0.081196  1.653427e+05  
std            1.421787      0.273143  9.831399e+04  
min            0.000000      0.000000  2.066700e+04  
25%            0.000000      0.000000  1.075070e+05  
50%            0.000000      0.000000  1.425940e+05  
75%            1.000000      0.000000  1.958420e+05  
max            4.000000      1.000000  2.265604e+06

#### Вывод:
***Столбец children***

Ошибок нет. Отрицательных значений нет. Минимальное количество детей - 0. Максимальное - 5.

***Столбец days_employed***

Отрицательных значений нет. Минимальное значение - 24 дня. Максимальное значение 401755 стажа - аномально. Допустим, что это значение в часах, тогда получим около 16740 дней стажа(что соответствует примерно 46 годам стажа). Учитывая, что максимальный возраст клиента 75 лет, предположение, что аномальные значения приведены в часах считаю верным. Приведем все аномальные значения в столбце к значениям в днях. Т.к. миниальный возраст для оформления по трудовому договору - 16 лет, а максимальное значение возраста - 75 лет. Возьмем границу стажа равную 59 лет (округлив получим 21500 дней). Все значения выше будем считать ошибочными и разделим их на 24 часа, приведя к дням.

***Столбец dob_years***

Минимальное значение возраста в годах 0 - т.е. клиенту нет года. Аномалия!

***Столбец total_income***

Максимальное значение достаточно большое, но аномальных значений нет.

In [26]:
# Приводим аномальные значения в столбце days_employed к значениям в днях
data.loc[data['days_employed'] > 21500, 'days_employed'] = data.loc[data['days_employed'] > 21500, 'days_employed'] / 24

In [27]:
# Проверим
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1        8437.00         42    высшее             0   
1         1        4024.00         36   среднее             1   
2         0        5623.00         33   среднее             1   
3         3        4124.00         32   среднее             1   
4         0       14177.75         53   среднее             1   
5         0         926.00         27    высшее             0   
6         0        2879.00         43    высшее             0   
7         0         152.00         50   среднее             1   
8         2        6929.00         35    высшее             0   
9         0        2188.00         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы  
5               покупка жилья                     B  операции с недвижимостью  
6           операции с жильем                     B  операции с недвижимостью  
7                 образование                     C     получение образования  
8       на проведение свадьбы                     C        проведение свадьбы  
9     покупка жилья для семьи                     C  операции с недвижимостью

In [28]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21331.000000   21331.000000  21331.000000  21331.000000   
mean       0.474005    4596.602298     43.277952      0.816980   
std        0.752378    5350.914876     12.575300      0.549282   
min        0.000000      24.000000      0.000000      0.000000   
25%        0.000000    1024.000000     33.000000      1.000000   
50%        0.000000    1996.000000     42.000000      1.000000   
75%        1.000000    5337.000000     53.000000      1.000000   
max        5.000000   18388.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21331.000000  21331.000000  2.133100e+04  
mean           0.974497      0.081196  1.653427e+05  
std            1.421787      0.273143  9.831399e+04  
min            0.000000      0.000000  2.066700e+04  
25%            0.000000      0.000000  1.075070e+05  
50%            0.000000      0.000000  1.425940e+05  
75%            1.000000      0.000000  1.958420e+05  
max            4.000000      1.000000  2.265604e+06

***Вывод: Максимальное значение в столбце days_employed равно 18388 дней, что соответствует 50 годам стажа.***

***Но у нас снова значения в столбце стали вещественными, т.к. мы применяли деление. (при делении целых чисел мы получим float). Преобразуем значения еще раз в целочисленные.***

In [29]:
data['days_employed'] = data['days_employed'].astype(int) # с помощью метода astype() преобразуем данные к типу int
data.head() # проверка

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

***Столбец dob_years***

In [30]:
# Посмотрим количество строк в датафрейме, где возраст меньше 18 лет(минимальны возраст заёмщика)
data.loc[data['dob_years'] < 18, 'dob_years'].count()

100

In [31]:
# Посмотрим, являются строки с возрастом менее 18 лет - нулём. Для этого посчитаем кол-во строк меньше 18 лет и не равных 0.
data.loc[(data['dob_years'] < 18) & (data['dob_years'] != 0), 'dob_years'].count()

0

***Вывод: Значение 0 говорит, что 100 строк содежат возраст заёмщика равный 0, что является ошибкой в данных. Вероятно, ошибка могла произойти вследствие заполнения нулями отсутствующих значений. Обработаем их как пропуски. Заменим, нулевые значения средними, которые посчитаем для каждого типа занятости(income_type).***

In [32]:
# Пройдемся циклом for по уникальным значениям колонки тип занятости
# Найдем среднее значение
# Заполним пропуски полученными значениями
for t in data['income_type'].unique():
    mean_value = data.loc[data['income_type'] == t, 'dob_years'].mean()
    data.loc[(data['dob_years'] == 0) & (data['income_type'] == t), 'dob_years'] = mean_value

In [33]:
data.loc[data['dob_years'] == 0, 'dob_years'].count()

0

***Вывод: Нулевые значения в колонке*** *dob_years* ***заменили средними по типу занятости.***

***Столбец education***

In [34]:
# Смотрим на значения столбца education
data['education'].sort_values().value_counts()

среднее                15075
высшее                  5227
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

***Вывод: данные впорядке***

***Столбец gender***

In [35]:
# смотрим значения
data['gender'].sort_values().value_counts()

F      14092
M       7238
XNA        1
Name: gender, dtype: int64

***Вывод: В столбце gender есть 1 некорректное значение***

In [36]:
# Посмотрим на строку с некорректным значением в столбце gender
data.loc[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0           2358       24.0  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

       total_income               purpose total_income_category  \
10701        203905  покупка недвижимости                     B   

               purpose_category  
10701  операции с недвижимостью

***Вывод: определить пол не представляется возможным. Т.к. столбец gender для исследований нам не понадобится. Оставим это значение как есть, чтобы не нарушать данные.***

### Ответы на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для ответа на этот вопрос сделаем следующее:
1. Cоздадим дополнительный столбец в датафрейме, в котором сделаем категоризацию по кол-ву детей.
2. Cоздадим сводную таблицу.

***Создадим функцию, которая будет выводить категорию людей по количеству детей:***
- ***нет детей***, если детей нет
- ***есть дети***, если 1 или 2 ребенка
- ***многодетный***, если 3 и более детей

In [37]:
# Функция категоризации по детям
def children_category(children):
    if 1 <= children <= 2:
        return 'есть дети'
    if children >= 3:
        return 'многодетный'
    return 'нет детей'

In [38]:
# Создаем столбец children_rank
data['children_rank'] = data['children'].apply(children_category)

In [39]:
# Проверяем результат. Выведем последние 5 значений
data.tail()

children  days_employed  dob_years education  education_id  \
21520         1           4529       43.0   среднее             1   
21521         0          14330       67.0   среднее             1   
21522         1           2113       38.0   среднее             1   
21523         3           3112       38.0   среднее             1   
21524         2           1984       40.0   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category children_rank  
21520  операции с недвижимостью     есть дети  
21521    операции с автомобилем     нет детей  
21522  операции с недвижимостью     есть дети  
21523    операции с автомобилем   многодетный  
21524    операции с автомобилем     есть дети

***Создадим сводную таблицу***

In [40]:
# Создаем таблицу методом pivot_table()
debt_children = data.pivot_table(index='children_rank', columns='debt', values='gender', aggfunc='count')
# Создаем столбцы
debt_children.columns = ['no_debt', 'debt']
# Считаем долю должников
debt_children['share_of_debtors'] = debt_children['debt'] / (debt_children['debt'] + debt_children['no_debt'])
# Переводим значения в столбце доля дожников в проценты
debt_children['share_of_debtors'] = debt_children['share_of_debtors'].map('{:.1%}'.format)
# Сортируем по столбцу доля лолжников по убыванию
debt_children.sort_values(by='share_of_debtors', ascending=False)

no_debt  debt share_of_debtors
children_rank                                
есть дети         6222   638             9.3%
многодетный        349    31             8.2%
нет детей        13028  1063             7.5%

***Вывод: Наличие детей повышает вероятность задолжености.*** Вероятно, это связано с дополнительными расходами на детей. Однако, можно заметить, что в многодетных семьях уровень должников ниже (8.2%), чем в семьях с одним или двумя детьми (9.3%). Это может быть связано либо с тем, что существуют программы поддержки многодетных семей или количество заемщиков в категории "многодетный" для данного сравнения недостаточно. Для категории граждан без детей самый низкий показатель должников (7.3%).

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

***Создадим сводную таблицу***

In [41]:
# Создаем таблицу методом pivot_table()
debt_family_status = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')
# Создаем столбцы
debt_family_status.columns = ['no_debt', 'debt']
# Считаем долю должников
debt_family_status['share_of_debtors'] = debt_family_status['debt'] / (debt_family_status['debt'] + debt_family_status['no_debt'])
# Переводим значения в столбце доля дожников в проценты
debt_family_status['share_of_debtors'] = debt_family_status['share_of_debtors'].map('{:.1%}'.format)
# Сортируем по столбцу доля лолжников по убыванию
debt_family_status.sort_values(by='share_of_debtors', ascending=False)

no_debt  debt share_of_debtors
family_status                                        
Не женат / не замужем     2523   273             9.8%
гражданский брак          3749   385             9.3%
женат / замужем          11334   927             7.6%
в разводе                 1105    84             7.1%
вдовец / вдова             888    63             6.6%

**Вывод:** ***Холостые и живущие в граждансокм браке пары больше подвержены риску стать должниками.*** Вероятно, это связано с тем, что у них больше свободы действий по распоряжению своими средствами. У них нет совместных накоплений, каждый распоряжается своими средствами по своему. Остальные категории, побывавшие в браке или на текущей момент в браке имеют меньшие проценты должников. Наиболее низкий процент у категории вдовец/вдова (6.6%). Это может быть связано с дополнительными средствами, доставшимися по наследству.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

***Запишем для наглядности категоризацию по уровню дохода, которую ранее упоминали в проекте.***

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

***Создадим сводную таблицу***

In [42]:
# Создаем таблицу методом pivot()
debt_total_income = data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')
# Создаем столбцы
debt_total_income.columns = ['no_debt', 'debt']
# Считаем долю должников
debt_total_income['share_of_debtors'] = debt_total_income['debt'] / (debt_total_income['debt'] + debt_total_income['no_debt'])
# Переводим значения в столбце доля дожников в проценты
debt_total_income['share_of_debtors'] = debt_total_income['share_of_debtors'].map('{:.1%}'.format)
# Сортируем по столбцу доля лолжников по убыванию
debt_total_income.sort_values(by='share_of_debtors', ascending=False)

no_debt  debt share_of_debtors
total_income_category                                
E                           20     2             9.1%
C                        14568  1353             8.5%
A                           23     2             8.0%
B                         4660   354             7.1%
D                          328    21             6.0%

***Вывод: Категория граждан (С) со средним доходом (50001–200000) больше подвержены риску стать должниками.*** Несмотря на то, что в итоговой таблицы должников данная категория граждан имеет не самый высокий процент по должникам, для точного исследования мы должны рассматривать репрезентативные выборки(категории граждан). Соответственно категории "E" и "A" мы не берем в расчет (из-за маленькой выборки), а выводы делаем опираясь на категории C, B, D. Сравнивая эти категории делаем вывод, что наибольший риск стать должниками у категории C, а наименьший у категории D.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

***Создадим сводную таблицу***

In [43]:
# Создаем таблицу методом pivot_table()
debt_purpose_category = data.pivot_table(index='purpose_category', columns='debt', values='gender', aggfunc='count')
# Создаем столбцы
debt_purpose_category.columns = ['no_debt', 'debt']
# Считаем долю должников
debt_purpose_category['share_of_debtors'] = debt_purpose_category['debt'] / (debt_purpose_category['debt'] + debt_purpose_category['no_debt'])
# Переводим значения в столбце доля дожников в проценты
debt_purpose_category['share_of_debtors'] = debt_purpose_category['share_of_debtors'].map('{:.1%}'.format)
# Сортируем по столбцу доля лолжников по убыванию
debt_purpose_category.sort_values(by='share_of_debtors', ascending=False)

no_debt  debt share_of_debtors
purpose_category                                        
операции с автомобилем       3879   400             9.3%
получение образования        3619   369             9.3%
проведение свадьбы           2130   183             7.9%
операции с недвижимостью     9971   780             7.3%

**Вывод:**  ***Граждане, берущие кредиты на получение образования и операции с автомобилем больше подвержены риску стать должниками.*** Эти две категории имеют одинаковый процент должников (9.3%). Граждане, которые берут кредит для проведения операций с недвижимостью более отвественно подходят к его погашению.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причинами могут быть: неверный ввод данных, сокрытие информации, повреждение файла данных при сохранении, технические проблемы.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Возможно потому что медиана устойчива к выбросам и аномальным значениям.

### Общий вывод.

1. 
- Наличие детей повышает вероятность задолжености.
- Холостые и живущие в граждансокм браке пары больше подвержены риску стать должниками. 
- Категория граждан со средним доходом (50001–200000) больше подвержены риску стать должниками.
- Граждане, берущие кредиты на получение образования и операции с автомобилем больше подвержены риску стать должниками.

2. 
- Причинами появления пропусков в данных могут быть: неверный ввод данных, сокрытие информации, повреждение файла данных при сохранении, технические проблемы.
- Заполнение пропусков для количественных переменных медианой является лучшим решением из-за устойчивости медианы к выбросам и аномальным значениям.